### PR TITLE
5: Add possibility to abort fetch requests

### PR DESCRIPTION
**Pull request by mrfrost91**

### CODE REVIEW

File: src/hooks/useNewsApi.ts [hunk 1]
```diff
@@ -36,37 +37,38 @@ const useNewsApi = () => {
...
-      setLoading(false);
-    }
-  }, [source, search, handleEmptySearchParams, state]);
+      try {
+        await apiToUse.fetchNews(search, signal);
+        setError(null);
+      } catch (err: unknown) {
+        setError(err as Error);
+      } finally {
+        if (shouldCreateNewInstance) api.current = apiToUse;
```
Comment (Rasalomkin) [inline src/hooks/useNewsApi.ts]: Curly braces here would make this more readable

File: src/hooks/useNewsApi.ts [hunk 2]
```diff
@@ -36,37 +37,38 @@ const useNewsApi = () => {
...
-  const fetchNewsApi = useCallback(async () => {
-    setLoading(true);
+  const fetchNewsApi = useCallback(
+    async (signal: AbortSignal) => {
+      setLoading(true);
 
-    const shouldCreateNewInstance = api.current.apiSlug !== source;
-    const apiToUse = shouldCreateNewInstance ? ApiFactory.createApi(source) : api.current;
+      const shouldCreateNewInstance = api.current.apiSlug !== source;
+      const apiToUse = shouldCreateNewInstance ? ApiFactory.createApi(source) : api.current;
```
Comment (Rasalomkin) [inline src/hooks/useNewsApi.ts]: I think logic around creating new api instance does not belong here.

File: src/hooks/useNewsApi.ts [hunk 3]
```diff
@@ -36,37 +37,38 @@ const useNewsApi = () => {
...
+        setError(null);
+      } catch (err: unknown) {
+        setError(err as Error);
+      } finally {
+        if (shouldCreateNewInstance) api.current = apiToUse;
+        setLoading(false);
+      }
+    },
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [source, search, handleEmptySearchParams, state, trigger],
```
Comment (Rasalomkin) [inline src/hooks/useNewsApi.ts]: I'm not sure if you need all of these as dependencies

File: src/hooks/useAbortController.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import { useEffect } from 'react';
+
+const useAbortController = (callback: (signal: AbortSignal) => Promise<void>): void => {
+  useEffect(() => {
+    const abortController = new AbortController();
+    const { signal } = abortController;
+
+    // eslint-disable-next-line no-void
+    void callback(signal);
```
Comment (Rasalomkin) [inline src/hooks/useAbortController.ts:9]: promise from callback is ignored here, is that intended? Also I'm not sure what the purpose of `void` here, won't it work without it?

File: src/hooks/useNewsApi.ts (hunk 1 above)
Comment (mrfrost91) [reply src/hooks/useNewsApi.ts]: Fixed

File: src/hooks/useNewsApi.ts (hunk 3 above)
Comment (mrfrost91) [reply src/hooks/useNewsApi.ts]: ESLint keeps warning me if I remove one of these dependencies, except trigger.

File: src/hooks/useAbortController.ts (hunk 4 above)
Comment (mrfrost91) [reply src/hooks/useAbortController.ts:9]: Yes, this was made on purpose.
ESLint suggests to use void in front of callbacks with ignored promises.

File: src/hooks/useNewsApi.ts (hunk 3 above)
Comment (mrfrost91) [reply src/hooks/useNewsApi.ts]: Removed a lot of dependencies here after refactoring

File: src/hooks/useNewsApi.ts (hunk 2 above)
Comment (mrfrost91) [reply src/hooks/useNewsApi.ts]: Currently I have no ideas how to implement the change of the api differently =(

**Pull request merged**